File: src/main/java/komu/blunt/asm/VM.java
package komu.blunt.asm;

import komu.blunt.asm.opcodes.OpCode;
import komu.blunt.eval.Environment;
import komu.blunt.eval.RootEnvironment;

import java.util.ArrayDeque;
import java.util.Deque;

import static com.google.common.base.Preconditions.checkNotNull;

public final class VM {
    
    public Object val;
    public Object env;
    public Object procedure;
    public Object arg;
    public int pc = 0;
    private final Instructions instructions;
    private final Deque<Object> stack = new ArrayDeque<>(4096);
    private final RootEnvironment globalEnvironment;
    public long steps = 0;

    public VM(Instructions instructions, Environment env, RootEnvironment globalEnvironment) {
        this.instructions = checkNotNull(instructions);
        this.globalEnvironment = checkNotNull(globalEnvironment);

        this.env = checkNotNull(env);
    }

    public Object run() {
        while (true) {
            if (pc >= instructions.count()) break;
            OpCode op = instructions.get(pc++);
            steps++;
            op.execute(this);
        }
        
        return val;
    }

    public Object get(Register register) {
        switch (register) {
        case VAL: return val;
        case ARG: return arg;
        case ENV: return env;
        case PC: return pc;
        case PROCEDURE: return procedure;
        }

        throw new IllegalArgumentException("unknown register: " + register);
    }
    
    public void set(Register register, Object value) {
        switch (register) {
        case VAL:
            val = value;
            break;
        case ARG:
            arg = value;
            break;
        case ENV:
            env = value;
            break;
        case PC:
            pc = (Integer) value;
            break;
        case PROCEDURE:
            procedure = value;
            break;
        default:
            throw new IllegalArgumentException("unknown register: " + register);
        }
    }
    
    public void jump(Label label) {
        this.pc = label.getAddress();
    }
    
    public void push(Object object) {
        stack.add(object);
    }
    
    public Object pop() {
        return stack.removeLast();
    }

    public void save(Register... registers) {
        for (Register register : registers)
            push(get(register));
    }
    
    public void restore(Register... registers) {
        for (int i = registers.length-1; i >= 0; i--)
            set(registers[i], pop());
    }

    public RootEnvironment getGlobalEnvironment() {
        return globalEnvironment;
    }
}

File: src/main/kotlin/komu/blunt/asm/VM.kt
package komu.blunt.asm

import komu.blunt.eval.Environment
import komu.blunt.eval.RootEnvironment
import komu.blunt.objects.Procedure
import java.util.*

class VM(private val instructions: Instructions, var env: Environment, val globalEnvironment: RootEnvironment) {

    var value: Any? = null
    var procedure: Procedure? = null
    var arg: Any? = null
    var pc = 0
    var steps = 0.toLong()

    private val stack = ArrayList<Any?>(4096)

    fun run(): Any? {
        while (true) {
            if (pc >= instructions.count) break
            val op = instructions[pc++]
            steps++
            op.execute(this)
        }

        return value
    }

    operator fun get(register: Register): Any? =
        when (register) {
            Register.VAL       -> value
            Register.ARG       -> arg
            Register.ENV       -> env
            Register.PC        -> pc
            Register.PROCEDURE -> procedure
        }

    operator fun set(register: Register, value: Any?) {
        when (register) {
            Register.VAL       -> this.value = value
            Register.ARG       -> arg = value
            Register.ENV       -> env = value as Environment
            Register.PC        -> pc = value as Int
            Register.PROCEDURE -> procedure = value as Procedure?
        }
    }

    fun push(value: Any?) {
        stack.add(value)
    }

    fun pop(): Any? =
        stack.removeAt(stack.lastIndex)
}
